src/dmd/sideeffect.d

```diff
@@ -211,14 +211,67 @@ private bool lambdaHasSideEffect(Expression e)
     return false;
 }
 
+private bool isAssignment(Expression e)
+{
+    extern (C++) final class IsAssignment : SemanticTimePermissiveVisitor
+    {
+        alias visit = typeof(super).visit;
+        bool result;
+
+        override void visit(AssignExp)
+        {
+            result = true;
+        }
+
+        override void visit(BinAssignExp)
+        {
+            result = true;
+        }
+    }
+
+    scope IsAssignment v = new IsAssignment();
+    e.accept(v);
+    return v.result;
+}
```

Comment:
This function would be more efficient if it would simply be replaced by an if statement:

```d
private bool isAssignmentExp(Expression e) @safe pure @nogc nothrow
{
    return e.isAssignExp || e.isBinAssignExp;
}